core/lib/Cardano/Address/Style/Shelley.hs

```diff
@@ -537,163 +560,245 @@ inspectShelleyAddress
-                0b11100000 | BS.length rest == credentialHashSize ->
-                    pure $ Json.object
-                        [ "address_style"     .= Json.String "Shelley"
-                        , "stake_reference"   .= Json.String "by value"
-                        , "stake_key_hash"    .= base16 fstHash
-                        , "stake_key_hash_bech32" .= bech32Stake fstHash
-                        , "network_tag"       .= network
-                        ]
-               -- 1111: reward account: scripthash32
-                0b11110000 | BS.length rest == credentialHashSize ->
-                    pure $ Json.object
-                        [ "address_style"     .= Json.String "Shelley"
-                        , "stake_reference"   .= Json.String "by value"
-                        , "script_hash"       .= base16 fstHash
-                        , "script_hash_bech32" .= bech32Script fstHash
-                        , "network_tag"       .= network
-                        ]
-                _ -> throwM UnknownAddrType
+-- | Determines whether an 'Address' is a valid address for the Cardano Shelley
+-- era. Shelley format addresses, as well as old-style Byron and Icarus
+-- addresses can be parsed by this function.
+--
+-- Returns either details about the 'Address', or 'ErrInspectAddress' if it's
+-- not a valid address.
+--
+-- @since 3.4.0
+eitherInspectAddress
+    :: Maybe XPub
+    -> Address
+    -> Either ErrInspectAddress InspectAddress
+eitherInspectAddress mRootPub addr = unpackAddress addr >>= parseInfo
   where
-    bytes  = unAddress addr
-    base16 = T.unpack . T.decodeUtf8 . encode EBase16
-
-    bech32With hrp = T.decodeUtf8 . encode (EBech32 hrp)
-    bech32Spending = bech32With CIP5.addr_vkh
-    bech32Stake = bech32With CIP5.stake_vkh
-    bech32Script = bech32With CIP5.script_vkh
-
-    ptrToJSON :: ChainPointer -> Json.Value
-    ptrToJSON ChainPointer{slotNum,transactionIndex,outputIndex} = Json.object
-        [ "slot_num" .= slotNum
-        , "transaction_index" .= transactionIndex
-        , "output_index" .= outputIndex
-        ]
+    parseInfo :: AddressParts -> Either ErrInspectAddress InspectAddress
+    parseInfo parts = case addrType parts of
+        -- 1000: byron address
+        0b10000000 ->
+            (bimap ErrIcarus InspectAddressIcarus (Icarus.eitherInspectAddress addr))
+            `orElse`
+            (bimap ErrByron InspectAddressByron (Byron.eitherInspectAddress mRootPub addr))
+        -- Anything else: shelley address
+        _ -> bimap ErrShelley InspectAddressShelley (parseAddressInfoShelley parts)
+
+-- | Returns either details about the 'Address', or
+-- 'ErrInspectAddressOnlyShelley' if it's not a valid Shelley address.
+parseAddressInfoShelley :: AddressParts -> Either ErrInspectAddressOnlyShelley AddressInfo
+parseAddressInfoShelley AddressParts{..} = case addrType of
+    -- 0000: base address: keyhash28,keyhash28
+    -- fixme: 28 byte hashes ????
```

Comment:
:thinking: hmmm ?

Reply:
Yeah those `scripthash32` comments were out of sync with the code and `shelley.cddl`.